test/Hessian_test.cpp

```diff
@@ -0,0 +1,171 @@
+{
+    T c = exp(x[0]);
+    T d = sin(x[1]);
+    T e = cos(x[2]);
+    return c + d + e;
+}
+
+template <class T>
+T single(std::vector<T> &x)
+{
+    return x[0] * x[0] * x[0] + x[0];
+}
+
+TEST(HessianTest, QuadraticForwardAdjoint)
+{
+    typedef xad::fwd_adj<double> mode;
+    typedef mode::active_type AD;
+    typedef mode::tape_type tape_type;
+
+    tape_type tape;
+
+    std::vector<AD> x = {3, 2};
+    xad::Hessian<AD> hes(quad<AD>, x, &tape);
+
+    std::vector<std::vector<AD>> cross_hessian = {{2.0, 0.0}, {0.0, 2.0}};
+    std::vector<std::vector<AD>> computed_hessian = hes.get();
+
+    for (unsigned int i = 0; i < cross_hessian.size(); i++)
+        for (unsigned int j = 0; j < cross_hessian.size(); j++)
+            ASSERT_EQ(cross_hessian[i][j], computed_hessian[i][j]);
+}
+
+TEST(HessianTest, QuadraticForwardAdjointWithIterator)
+{
+    typedef xad::fwd_adj<double> mode;
+    typedef mode::active_type AD;
+    typedef mode::tape_type tape_type;
+
+    tape_type tape;
+
+    std::vector<AD> x = {3, 2};
+    std::vector<std::vector<AD>> computed_hessian(x.size(), std::vector<AD>(x.size(), 0.0));
+    xad::Hessian<AD> hes(quad<AD>, x, &tape, computed_hessian.begin(), computed_hessian.end());
+
+    std::vector<std::vector<AD>> cross_hessian = {{2.0, 0.0}, {0.0, 2.0}};
+
+    for (unsigned int i = 0; i < cross_hessian.size(); i++)
+        for (unsigned int j = 0; j < cross_hessian.size(); j++)
+            ASSERT_EQ(cross_hessian[i][j], computed_hessian[i][j]);
+}
+
+TEST(HessianTest, SingleInputForwardAdjoint)
+{
+    typedef xad::fwd_adj<double> mode;
+    typedef mode::active_type AD;
+    typedef mode::tape_type tape_type;
+
+    tape_type tape;
+
+    std::vector<AD> x = {3};
+    xad::Hessian<AD> hes(single<AD>, x, &tape);
+
+    std::vector<std::vector<AD>> cross_hessian = {{18.0}};
+    std::vector<std::vector<AD>> computed_hessian = hes.get();
+
+    for (unsigned int i = 0; i < cross_hessian.size(); i++)
+        for (unsigned int j = 0; j < cross_hessian.size(); j++)
+            ASSERT_EQ(cross_hessian[i][j], computed_hessian[i][j]);
+}
+
+TEST(HessianTest, QuadraticForwardForward)
+{
+    typedef xad::fwd_fwd<double> mode;
+    typedef mode::active_type AD;
+
+    std::vector<AD> x = {3, 2};
+    xad::Hessian<AD> hes(quad<AD>, x);
+
+    std::vector<std::vector<AD>> cross_hessian = {{2.0, 0.0}, {0.0, 2.0}};
+    std::vector<std::vector<AD>> computed_hessian = hes.get();
+
+    for (unsigned int i = 0; i < cross_hessian.size(); i++)
+        for (unsigned int j = 0; j < cross_hessian.size(); j++)
+            ASSERT_EQ(cross_hessian[i][j], computed_hessian[i][j]);
+}
+
+TEST(HessianTest, QuadraticForwardForwardWithIterator)
+{
+    typedef xad::fwd_fwd<double> mode;
+    typedef mode::active_type AD;
+
+    std::vector<AD> x = {3, 2};
+    std::vector<std::vector<AD>> computed_hessian(x.size(), std::vector<AD>(x.size(), 0.0));
+    xad::Hessian<AD> hes(quad<AD>, x, computed_hessian.begin(), computed_hessian.end());
```

Comment:
The way you call them is completely functional (which is good). There's actually no need for a class instance (can be used internally if desired, in `xad::detail` namespace). I think functional makes a more straightforward interface:

```c++
compute_hessian(quad<AD>, x, begin(computed_hessian), end(computed_hessian));
```

or for the version without iterator:

```c++
auto computed_hessian = compute_hessian(quad<AD>, x);
```

Reply:
If I'm understanding correctly:
- Remove Hessian class
- Implement its methods in xad::detail

If that's the case, I'll do that after I resolve the other open discussion.

Reply:
You can put into `detail` whatever you like / what makes it easer to solve the problem without too much repetition. It could be a set of helper functions, it could also be the whole class if that helps.